src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -718,10 +720,64 @@ InterpreterSelectQuery::InterpreterSelectQuery(
     sanitizeBlock(result_header, true);
 }
 
+void InterpreterSelectQuery::executePutInCache(QueryPlan & query_plan)
+{
+    auto settings = context->getSettingsRef();
+    QueryCachePtr query_cache = context->getQueryCache();
+    if (!settings.query_cache_active_usage || query_cache == nullptr)
+    {
+        return;
+    }
+
+    auto query_cache_key = CacheKey{query_ptr, result_header, settings,
+                                    settings.share_query_cache
+                                    ? std::optional<String>()
+                                    : std::make_optional<String>(context->getUserName())
+                                    };
+    size_t num_query_runs = query_cache->recordQueryRun(query_cache_key);
+    if (query_cache->containsResult(query_cache_key) || num_query_runs < settings.min_query_runs_before_caching)
+    {
+        return;
+    }
+
+    auto caching_step = std::make_unique<CachingStep>(query_plan.getCurrentDataStream(), context->getQueryCache(), query_cache_key);
+    caching_step->setStepDescription("Put query result in cache");
+    query_plan.addStep(std::move(caching_step));
+}
+
+bool InterpreterSelectQuery::executeReadFromCache(QueryPlan & query_plan)
```

Comment:
What happens if the underlying data changed? That should invalidate the result cache, right?